src/sage/matrix/matrix_polynomial_dense.pyx

```diff
@@ -4079,3 +4084,462 @@ cdef class Matrix_polynomial_dense(Matrix_generic_dense):
+        completed into a basis of `\Bold{K}[x]^{n}` (or equivalently, `A` has
+        unimodular column bases, or also, if the rows of `A` generate all
+        polynomial vectors in the rational row space of `A`), then `C` provides
+        such a completion. In this case, `[[A],[C]]` is unimodular: it is
+        invertible over `\Bold{K}[x]`, and `det([[A],[C]])` is a nonzero
+        element of the base field `\Bold{K}`.
+
+        INPUT:
+
+        - ``row_wise`` -- (optional, default: ``True``) boolean, if ``True`` then
+          compute a row-wise completion, else compute a column-wise completion.
+
+        - ``algorithm`` -- (optional, default: ``approximant``) selects the
+          approach for computing the completion; currently supported:
+          ``approximant`` and ``smith``.
+
+        OUTPUT: a matrix over the same base ring as the input matrix, which forms a
+        completion as defined above.
+
+        ALGORITHM:
+
+        - ``approximant``: the approximant-based algorithm follows the ideas in
+          [ZL2014]_ , based on polynomial reversals combined with the
+          computation of a minimal kernel basis and a minimal approximant
+          basis.
+
+        - ``smith``: the Smith form-based algorithm computes the Smith form of
+          this matrix along with corresponding unimodular transformations, from
+          which a completion is readily obtained.
+
+        EXAMPLES:
+
+        Three polynomials whose GCD is `1` can be completed into a unimodular
+        matrix::
+
+            sage: ring.<x> = GF(7)[]
+            sage: mat = matrix([[x*(x-1)*(x-2), (x-2)*(x-3)*(x-4), (x-4)*(x-5)*(x-6)]])
+            sage: print(mat)
+            [      x^3 + 4*x^2 + 2*x   x^3 + 5*x^2 + 5*x + 4   x^3 + 6*x^2 + 4*x + 6]
+            sage: rcomp = mat.basis_completion(); print(rcomp)
+            [        5*x^2 + 4*x + 1             5*x^2 + 2*x                   5*x^2]
+            [          2*x^3 + 4*x^2 2*x^3 + 6*x^2 + 2*x + 1       2*x^3 + x^2 + 3*x]
+            sage: basis = mat.stack(rcomp); print(basis)
+            [      x^3 + 4*x^2 + 2*x   x^3 + 5*x^2 + 5*x + 4   x^3 + 6*x^2 + 4*x + 6]
+            [        5*x^2 + 4*x + 1             5*x^2 + 2*x                   5*x^2]
+            [          2*x^3 + 4*x^2 2*x^3 + 6*x^2 + 2*x + 1       2*x^3 + x^2 + 3*x]
+            sage: basis.determinant()
+            6
+
+        The following matrix has rank `2` and trivial Smith form. It can be
+        completed row-wise into a `3 \times 3` unimodular matrix (column-wise,
+        there is nothing to complete)::
+
+            sage: mat = matrix(ring, 2, 3, \
+                    [[x^2 + 5*x + 5,   3*x^2 + x + 3, 4*x^2 + 5*x + 4], \
+                     [5*x^2 + 4*x,   3*x^2 + 4*x + 5, 5*x^2 + 5*x + 3]])
+            sage: rcomp = mat.basis_completion(); print(rcomp)
+            [  2*x^2 + 1 4*x^2 + 3*x 2*x^2 + 3*x]
+            sage: mat.stack(rcomp).determinant()
```

Comment:
Perhaps I misunderstood something, but shouldn't a unimodular matrix have determinant $\pm 1$? Here $3 \not \equiv \pm 1 \pmod{7}$.

Also, it is a good idea to have an example over $\mathbb{Q}$ or a number field (IMO).

Reply:
I might be mistaken but I think the general definition is that a (square) matrix over some ring is said to be unimodular if it is invertible (over that ring), or equivalently that its determinant is a unit (in the ring). For univariate polynomials, units are all nonzero constants; the set of units being +-1 would be e.g. for relative integers.

Reply:
Ahh okay, you are correct, wikipedia is wrong (as always). I think I had this exact same question and discussion at least 5 times in the past year...Ahh okay, that makes sense, wikipedia is wrong (as always). I think I had this exact same question and discussion at least 5 times in the past year...

Reply:
examples added!